### ISSUE TITLE
Navigation button on a Place or Person page

### ISSUE THREAD

**Issue by estellecomment** [medic/medic-webapp#2803]
Right now 
 - there is no navigation button at the top of desktop RHS
 - there is a navigation button on mobile, but it always goes back to the place list.

Navigation button, description in design spec : https://docs.google.com/document/d/1UwSe8H3xykvOUwu65KxJYgBTxv40Y_9b3CPgRbcXOH0/edit#bookmark=id.qkrffmrtf0zt

Note : navigation when you are filling in a form from within the People tab is different, and described in #2763

**Comment by SCdF** [comment]
@diannakane I'm not sure I completely understand how this would work. There are only a couple of sentences on this in the linked document that I can see, so let me write some user flows of my own and you can clarify.

I've written these with the mobile UI in mind, but I am presuming the only distinction between the desktop view and the mobile view is that where the mobile view would show a back button to go back to the list view, the desktop view shows nothing (because you can already see the list view, and thus that button wouldn't make sense). Your clarification would be helpful here.

On to the scenarios. First, the two (hopefully) obvious ones:

Story 1:
 - In contact view
 - Taps on any contact
 - Taps back button
 - Is back on the contact view.

Story 2:
 - Is in contact view
 - Taps on their home area
 - Taps on a household
 - Taps back button
 - Is back looking at their home area, which is the **parent** (as described) of the household
 - Taps back button
 - Is back in the contact view

Now for complications:

Story 3:
 - Is in contact view
 - Taps on their home area
 - Taps on the "family members" (which is them)
 - Tap the back button
 - Is.... **where?** Back at the home area, which is **not** their parent? Or up one layer to their family, which **is** their parent?

^ This distinction describes a fracture in what this back button is supposed to represent. Is it tap history, like a standard back button? Or is it a hierarchy button?

To expand on that a little:

Story 3b:
 - Is in contact view
 - Taps on their home area
 - Taps on the "family members" (which is them)
 - Refreshes the page OR Takes the URL---which has the ID of them in it (the doc they are viewing)---and puts it in a new tab
 - ...? Do they see a back button at all (a traditional back button would have no history here)? If so, does it take them a) to the list view, b) to the home area, or c) to their parent place

To talk about the same problem from a different angle, it is also not clear which of these two goals we are trying to satisfy:
 - We want to always show contacts in relation to their parents and their children (the latter being dealt with already, ie households show family members etc)
 - We want the on-screen mobile navigation button to work like a traditional browser back button, where each different "thing" you look at goes on the history stack and each tap of the back button pops the previous thing off that stack

AFAICT these two things conflate and break each other in various scenarios, which is where my confusion is coming from.

FWIW I'm not sure if it's a bug or intentional that on mobile the **hardware** back button doesn't go back correctly either. It certainly surprises me, as it does what you'd expect on desktop.

**Comment by sglangevin** [comment]
@SCdF I'm not sure when story 3 would be possible. How would you be able to click on a person from your home area, other than from the list of People, who are people whose parent is the area? In order to click on a family member, you have to be on a household/family profile page, right? And the household/family page is the parent page to the family member. This probably doesn't resolve the question in story 3b, but I'm curious how you would achieve the navigation that you've described.

The hardware back button has not really worked correctly at many points along the way so far, so I'd say that is probably a bug and it would be nice to have either the hardware back button perform the same function as the navigation back button or at least clearly describe how the two should work (if they need to work differently).

**Comment by SCdF** [comment]
@sglangevin oh you're right, if you are the primary contact (currently labeled "family member", we might need to fix that…) then you're placed under that thing in the hierarchy! That's cool.

So I think the work could be described as:
 - In desktop view show the contact's parent's name at the top of the contact view as a clickable link, which links to displaying the parent document in the contact view. Unless you're already top level, in which case show nothing
 - In mobile view fix both the hard and current soft back button so it follows normal conventions and goes back to the last thing you looked at. If you have no back button history (ie you refreshed) hitting back while viewing a contact goes back to the contact list.

For desktop we've actually already done this, it's currently displayed as "Belongs to" in the first section:
![image](https://cloud.githubusercontent.com/assets/583851/19133468/432a6a82-8b50-11e6-98b0-4b2a324b09c3.png)
We can click this and walk up to the area:
![image](https://cloud.githubusercontent.com/assets/583851/19133476/4c79077e-8b50-11e6-8d5f-0f958962be76.png)
And while we can see the branch, we can't click it because we're a restricted user and we don't have that document available locally.

As Admin we do have this data, so we can keep going:
![image](https://cloud.githubusercontent.com/assets/583851/19133505/78f80e08-8b50-11e6-99dd-9a761b16a070.png)
Which is the top, which has no parent:
![image](https://cloud.githubusercontent.com/assets/583851/19133519/8cd87642-8b50-11e6-86de-03bb3d06e771.png)

So if I've got this right, we basically want to duplicate that link at the top? Or move it? The screenshots in the google doc linked at the top of this ticket show both, but we can put that down to mockups not being perfect:
![image](https://cloud.githubusercontent.com/assets/583851/19133544/b767cbb0-8b50-11e6-9866-61be7ec77e84.png)
I think based on later parts of that document we want to move it, because we want to convert the current belongs to link to breadcrumbs, as shown in a later part of that same document:
![image](https://cloud.githubusercontent.com/assets/583851/19133563/d96db026-8b50-11e6-95c2-18a7818839c3.png)

Am I understanding that right? It seems a little weird because it seems like they do the same job: ie you have the same information in two places:
 - At the top you have a link to the direct parent
 - Just below that you all links to all parents, including the direct parent, in breadcrumb form

**Comment by sglangevin** [comment]
@SCdF I think the intention was to have a title at the top (e.g. Bwaise Branch in the first mock-up) as well as a back button. The back button would go up one level to that person or place's parent. The title is the title of the page you are looking at, not the page you will go to if you hit the back button (Bwaise Branch, the page title, should probably be farther away from the back button).

The breadcrumbs then show you the full hierarchy above the person or place you are viewing. I think the intention was to have all the breadcrumbs and the back button at the top.

I think the remaining question is what happens if you are viewing a person's profile, then click on the CHW Area in the breadcrumbs, then hit the back button? Do you go up a level of the hierarchy to the branch (let's assume you have permission to view branches)? Or do you go back to the person's profile that you came from?

**Comment by SCdF** [comment]
>  The title is the title of the page you are looking at

Are you saying in the Bwaise Branch screenshot above that the Dickton Mtanga's Area bit wouldn't actually be there? It's unfortunately quite confusing because the mockup is not internally consistent :-(.

> I think the remaining question is what happens if you are viewing a person's profile, then click on the CHW Area in the breadcrumbs, then hit the back button?

We have to be careful with language here. Your scenario applies to 4 different situations: a combination of either the mobile or desktop view, and the hardware back button or software back button (MH/MS/DH/DS).

**Comment by diannakane** [comment]
@estellecomment and I talked about the software back button being the immediate parent place, except when viewing a task or history record from within the People tab. In those scenarios, it's possible that you navigated to an individual's task or record from a family page, and would want to return back to the family page using back, or at the end of the form. Referenced here: https://docs.google.com/document/d/1UwSe8H3xykvOUwu65KxJYgBTxv40Y_9b3CPgRbcXOH0/edit#bookmark=id.907drg3i839q

The hardware back button would take you back to where you last were. These may oftentimes be the same thing, since most flows take you through a parent place first. In the scenario that Sharon described in the last comment, the software back button would take you to the parent place of the CHW area, and the HBB would take you to the person's profile. The same could happen if you click on a person in the LHS from a list of search results (the only way people will show up in the LHS). The software back button would show their parent place, and the HBB would take you back to the last thing you were viewing.

**Comment by diannakane** [comment]
@SCdF Bwaise Branch screenshot above is correct.

**Comment by estellecomment** [comment]
(hwoa much activity on this)
Dianna summarized this well : the software navigation button is not really a back button : it's a button to go up one level in hierarchy. 
The HBB is actually a back button. :) Like a browser back button.

Yes, the "Belongs to" field (which I've also called breadcrumbs, don't know if that's misleading) partially repeats the function of the navigation button. (except is has more levels) I raised that with @diannakane in the past, and remember that it's a known thing (but don't remember why!)

For this ticket : 
 - [ ] Desktop : add a navigation button, with the name of the place it's navigating to, at the top of the page. (screenshot below)
 - [ ] Mobile : the navigation button is already there but is going to the wrong place. Make it point to the parent place. (Ignore the text in the orange bar : it's not the text of the navigation button, it's the title of the page.)


Screenshot 1 of the page for "Dickson Mtanga's Area", with a navigation button going to its parent, "Bwaise Branch" 

![image](https://cloud.githubusercontent.com/assets/911434/19144607/e3d865b0-8baa-11e6-8dc4-980c1e463c5a.png)

For more examples check out [the mocks on Zeppelin](https://app.zeplin.io/project.html#pid=57d311edb055084578d006ec&dashboard).

**Comment by SCdF** [comment]
> Mobile : the navigation button is already there but is going to the wrong place. Make it point to the parent place. (Ignore the text in the orange bar : it's not the text of the navigation button, it's the title of the page.)

@estellecomment so to be clear, this means if you tap into a contact the **only way** to get back to the contact list, or in fact any other place at all (since the top bar gets hidden) is to hit the hardware back button enough times?

**Comment by SCdF** [comment]
Also, if my understanding is correct, may I make a suggestion that we don't use `<` as the iconography for this link?
 - It mimicks the iconography and placement of a browser back button
 - It mimicks back buttons in iOS apps, and various android apps and web apps that have aped that design language
 - As well as, I would imagine, a lot of design language about going "back"

Apologies if this is viewed differently by our users, who I admit are a completely different userbase to those I'm used to. Having said said, IMO `^` would be more appropriate.

**Comment by SCdF** [comment]
This ticket is blocked on a conversation between myself (or someone else if they want to work on the ticket, I'm not precious about it) and Dianna, where at the end of the call we have concrete answers, flows and rules that allow devs to implement this without requiring more questions.

Some thought experiments that might help on the call:
 - What is the flow, in both desktop and mobile, using both hardware and software back buttons, in the following situations
  - User starts on contact list view, clicks on a contact, and then goes back 
  - User starts on contact list view, clicks a contact, then clicks a deeper related contact (area->family say), then goes back
  - User starts on viewing a contact (ie after a refresh), and then clicks back (may only apply to mobile).

**Comment by diannakane** [comment]
Updated the spec with flows to these scenarios and have run it by @lesterng
who agrees with the decisions. Bookmark link in dev slack channel.

Sounds like the most confusing part is "starting from a contact list" is an
edge case only possible through search results.

**Comment by SCdF** [comment]
> "starting from a contact list" is an edge case only possible through search results.

What do you mean here? Isn't this the default (and in fact maybe only) way to view a contact? From the list? Maybe this is a terminology issue

**Comment by diannakane** [comment]
Sounds like a terminology issue. I interpret "contact list" to be a list of
people. Currently the way only to view people in the LHS is through
returning search results.

We have lists of people who belong to places in the RHS, but if you click
on any individual person, their page does not contain any sub-places and
does not link anywhere else.

**Comment by estellecomment** [comment]
I think : Contact = Place or Person
Contact list = the list in the LHS on desktop

**Comment by diannakane** [comment]
Got it. Can we use the language of People & Places to differentiate?

**Comment by SCdF** [comment]
@diannakane I don't understand what your terms are. When you say "the language of People & Places", it's not clear to me what that language is. Like I said higher up the thread, we should get on a call.I don't think playing github issue ping pong is going to be very efficient. We can lock down terms at the start of that call.

**Comment by SCdF** [comment]
@diannakane and I had a chat: turns out we're on the same page, and terminology and other things were throwing the discussion.

There are two requirements to solve: a hierarchy link; and back button rules.

**Thing 1**: on desktop, if the user is viewing a person or place, and has access to that person or place's parent (ie a CHW can access a family's area but not branch / dh, an admin can see everything), there will be a link in the top left hand corner of the right-hand-side view, that links to their parent, with the rough label `< PARENT_NAME`.

**Thing 2**: the back button goes back.

**Thing 2.5**: Thing 2 is true, **except** that on Android we also judiciously[1] apply Android design guidelines across the whole app, that may override that rule.

We did some of this in #1564. However, #1565 is still open (backlogged since Feb '15), and is for working through some of the more complex rules.

Specifically for this ticket, we're interested in navigation situations that make it more than one tap away from the people and places search pane (ie content that is normally in the LHS in desktop view). For example, searching for a family, tapping that family, and then tapping on a person in that family. At the moment, hitting back at any point on Android will take you back to the search pane, whereas it should in those situations simply take you back, like a browser.

[1] I say judicious because as someone who uses Android devices and apps written for them, they are very much guidelines and not rules.

**Comment by SCdF** [comment]
There has been a lot of conversation and clarification on this ticket. To make things cleaner I'm going to close this ticket and raise two others with specific instructions on each. This will allow us to discuss the two issues separately without conflating them.

**Comment by SCdF** [comment]
Added #2819 and #2820
